tests/integration/test_decompositions_integration.py

```diff
@@ -448,3 +448,111 @@ def test_rotated_squeezed(self, setup_eng, cutoff, hbar, tol):
+    def test_CXgate(self, setup_eng, pure, hbar, tol):
+        """Test the action of the CX gate in phase space"""
+        if not pure:
+            pytest.skip("Test only runs on pure states")
+        N = 2
+        eng, prog = setup_eng(N)
+        r = 3
+        x1 = 2
+        x2 = 1
+        p1 = 0.0
+        p2 = 0.0
+        s = 0.5
+        with prog.context as q:
+            ops.Sgate(r) | q[0]
+            ops.Xgate(x1) | q[0]
+            ops.Zgate(p1) | q[0]
+            ops.Sgate(r) | q[1]
+            ops.Xgate(x2) | q[1]
+            ops.Zgate(p2) | q[1]
+            ops.CXgate(s) | q
+        state = eng.run(prog).state
+        CXmat = np.array([[1, 0, 0, 0], [s, 1, 0, 0], [0, 0, 1, -s], [0, 0, 0, 1]])
+        Vexpected = 0.5 * hbar * CXmat @ np.diag(np.exp([-2 * r, -2 * r, 2 * r, 2 * r])) @ CXmat.T
+        # Checks the covariance matrix is transformed correctly
+        assert np.allclose(state.cov(), Vexpected, atol=tol)
+        rexpected = CXmat @ np.array([x1, x2, p1, p2])
+        # Checks the means are transformed correctly
+        assert np.allclose(state.means(), rexpected, atol=tol)
+
+    @pytest.mark.backends("gaussian")
+    def test_CZgate(self, setup_eng, pure, hbar, tol):
+        """Test the action of the CZ gate in phase space"""
+        if not pure:
+            pytest.skip("Test only runs on pure states")
+        N = 2
+        eng, prog = setup_eng(N)
+        r = 3
+        x1 = 2
+        x2 = 1
+        p1 = 0.0
+        p2 = 0.0
+        s = 0.5
+        with prog.context as q:
+            ops.Sgate(r) | q[0]
+            ops.Xgate(x1) | q[0]
+            ops.Zgate(p1) | q[0]
+            ops.Sgate(r) | q[1]
+            ops.Xgate(x2) | q[1]
+            ops.Zgate(p2) | q[1]
+            ops.CZgate(s) | q
+        state = eng.run(prog).state
+        CZmat = np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, s, 1, 0], [s, 0, 0, 1]])
+        Vexpected = 0.5 * hbar * CZmat @ np.diag(np.exp([-2 * r, -2 * r, 2 * r, 2 * r])) @ CZmat.T
+        # Checks the covariance matrix is transformed correctly
+        assert np.allclose(state.cov(), Vexpected, atol=tol)
+        rexpected = CZmat @ np.array([x1, x2, p1, p2])
+        # Checks the means are transformed correctly
+        assert np.allclose(state.means(), rexpected, atol=tol)
+
+    @pytest.mark.backends("fock")
+    def test_S2gate_fock(self, setup_eng, pure, hbar, tol):
+        """Test the action of the S2gate gate on vacuum in Fock space"""
+        if not pure:
+            pytest.skip("Test only runs on pure states")
+        N = 2
+        eng, prog = setup_eng(N)
+        nbar = 1
+        s = np.arcsinh(np.sqrt(1.0))
+        phi = 0 * np.pi / 3
+        with prog.context as q:
+            ops.S2gate(s, phi) | q
+        state = eng.run(prog).state
+        ket = state.ket()
+        n, _ = ket.shape
+        diag_elems = (1 / np.sqrt(1 + nbar)) * (
+            np.exp(1j * phi) * np.sqrt((nbar / (1 + nbar)))
+        ) ** (list(range(n)))
```

Comment:
I'm a bit confused why this line works; the LHS of the ** should be a Python float, while the RHS is a list - how come it evaluates as a NumPy array?

Reply:
Oh, it's because the LHS is of type `np.float64`.

Maybe it's worth being a bit explicit, and having the exponent be `np.arange(n)` instead of `list(range(n))`

Reply:
Cool! Will do. To be honest I was not sure why it worked.